lib/vdsm/virt/vmdevices/storage.py

```diff
@@ -923,22 +925,54 @@ def get_snapshot_xml(self, snap_info):
 
     def on_block_threshold(self, reported_path):
         """
-        Callback to be executed when we receive a BLOCK_THRESHOLD
-        event from libvirt. We mark this drive accordingly,
-        so the periodic check will pick up this drive for
-        extension.
+        Called when disk allocation excceeds configured threshold.
+
+        Mark this drive for extension, to be picked by the periodic monitor
+        later.
+
+        Return True if threshold state was modified.
         """
         with self._lock:
             if reported_path != self._path:
                 self.log.debug(
                     "block threshold event mismatch drive %r path=%r "
                     "reported path=%r - ignored",
                     self.name, self._path, reported_path)
-                return
+                return False
+
+            if self._threshold_state == BLOCK_THRESHOLD.EXCEEDED:
+                self.log.debug(
+                    "drive %r block threshold already exceeded, ignored",
+                    self.name)
+                return False
+
+            self.log.info("drive %r block threshold exceeded", self.name)
+            self._threshold_state = BLOCK_THRESHOLD.EXCEEDED
+            self.exceeded_time = time.monotonic_time()
+            return True
```

Comment:
These two blocks here and in on_enospc differ only in a single word in a log message.  Shouldn’t the common code be extracted?

Reply:
The shared code is 3 trivial lines and is we have only 2 copies, and is safe only when called under the lock.

We can do this:

    def _set_exceeded_locked(self):
        self._threshold_state = BLOCK_THRESHOLD.EXCEEDED
        self.exceeded_time = time.monotonic_time()

And both event handlers will do:

    log ...
    self._set_exceeded_locked()
    return True

And we need to hope that nobody will merge code calling _set_exceeded_locked() from other places, possibly without the lock.

Is this really better?

Reply:
If I’m looking correctly, there are actually 8-11 shared lines (depending on how you count them) + 1 line that differs only a single word.  This is worth of deduplication.

I don’t think there is a risk of musing a shared method but if you want to prevent it then the two methods may be merged and distinguished by an argument.

Reply:
I'll improve this in the next version.